File: src/main/java/kr/co/bookand/backend/account/domain/Account.java
package kr.co.bookand.backend.account.domain;

import com.fasterxml.jackson.annotation.JsonIgnore;
import kr.co.bookand.backend.account.domain.dto.AccountDto;

import kr.co.bookand.backend.bookmark.domain.Bookmark;
import kr.co.bookand.backend.bookstore.domain.ReportBookStore;
import kr.co.bookand.backend.common.domain.BaseEntity;
import kr.co.bookand.backend.feedback.domain.Feedback;
import lombok.*;
import org.springframework.format.annotation.DateTimeFormat;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@AllArgsConstructor
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Builder
public class Account extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "account_id")
    private Long id;

    private String email;
    private String nickname;
    private String password;
    private String provider;
    private String providerEmail;
    private String profileImage;

    @DateTimeFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss")
    private LocalDateTime lastLoginDate;

    @DateTimeFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss")
    private LocalDateTime signUpDate;

    @Enumerated(EnumType.STRING)
    private Role role;

    @JsonIgnore
    @OneToMany(mappedBy = "account", cascade = CascadeType.ALL)
    private List<Feedback> feedbackList = new ArrayList<>();

    @JsonIgnore
    @OneToMany(mappedBy = "account", cascade = CascadeType.ALL)
    private List<Bookmark> bookmarkList = new ArrayList<>();

    @JsonIgnore
    @OneToMany(mappedBy = "account", cascade = CascadeType.ALL)
    private List<ReportBookStore> tipList = new ArrayList<>();

    public AccountDto.MemberRequest toAccountRequestDto(String suffix) {
        return new AccountDto.MemberRequest(email, email + suffix, nickname, provider);
    }

    public void updateNickname(String nickname) {
        this.nickname = nickname;
    }

    public void setIdForTest(Long id) {
        this.id = id;
    }

    public void updateBookmarkList(List<Bookmark> bookmark) {
        if (bookmarkList == null) {
            bookmarkList = new ArrayList<>();
        }
        bookmarkList.addAll(bookmark);
    }

    public void updateProfileImage(String profileImage) {
        this.profileImage = profileImage;
    }

    public void updateLastLoginDate(LocalDateTime lastLoginDate) {
        this.lastLoginDate = lastLoginDate;
    }

    public void updateSignUpDate(LocalDateTime signUpDate) {
        this.signUpDate = signUpDate;
    }
}

File: src/main/kotlin/kr/co/bookand/backend/account/model/Account.kt
package kr.co.bookand.backend.account.model

import kr.co.bookand.backend.bookmark.model.Bookmark
import kr.co.bookand.backend.bookstore.model.ReportBookstore
import kr.co.bookand.backend.common.model.BaseEntity
import kr.co.bookand.backend.feedback.model.Feedback
import java.time.LocalDateTime
import javax.persistence.*

@Entity
class Account(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "account_id")
    var id: Long = 0,

    var email: String,
    var password: String,
    var nickname: String,
    var provider: String,
    var providerEmail: String,
    var profileImage: String? = null,
    var lastLoginDate: LocalDateTime? = null,
    var signUpDate: LocalDateTime? = null,

    @Enumerated(EnumType.STRING)
    var role: Role,
    @Enumerated(EnumType.STRING)
    var accountStatus: AccountStatus,

    @OneToMany(mappedBy = "account", cascade = [CascadeType.ALL], orphanRemoval = true)
    val feedbackList: MutableList<Feedback> = mutableListOf(),

    @OneToMany(mappedBy = "account", cascade = [CascadeType.ALL], orphanRemoval = true)
    var bookmarkList: MutableList<Bookmark> = mutableListOf(),

    @OneToMany(mappedBy = "account", cascade = [CascadeType.ALL], orphanRemoval = true)
    var tipList: MutableList<ReportBookstore> = mutableListOf(),

    ) : BaseEntity() {
    fun updateProfile(profileImage: String, nickname: String) {
        this.profileImage = profileImage
        this.nickname = nickname
    }

    fun updateLastLoginDate() {
        this.lastLoginDate = LocalDateTime.now()
    }

    fun updateAccountStatus(accountStatus: AccountStatus) {
        this.accountStatus = accountStatus
    }

    fun updateBookmarks(bookmarkList: MutableList<Bookmark>) {
        this.bookmarkList = bookmarkList
    }

    fun deletedBanAccount() {
        this.nickname = "BAN"
        this.password = "BAN"
        this.email = "BAN"
        this.provider = "BAN"
        this.providerEmail = "BAN"
        this.profileImage = "BAN"
        this.role = Role.SUSPENDED
    }

}